docs/design/proposals/VRG-Reconciliation.md

```diff
@@ -0,0 +1,132 @@
+    1. If not bound, continue without requeue
+        1. **NOTE** We need PVC to be in deleted state, to prevent use by pods
+        when marking it as secondary. If it is still not bound and is deleted,
+        it will never be bound. If not deleted we cannot mark it secondary
+        anyway. So we skip without requeuing for this PVC.
+    1. Check `vr-protected` finalizer,
+        1. If found, check and remove (stale) `vr-protecting` finalizer
+        1. If not found, continue no requeue
+            1. **NOTE** If being deleted then we cannot add our finalizers as
+            we enable it to be replication protected, if not being deleted
+            then we cannot ensure use by pods as we setup volume as secondary.
+            Hence, we skip without requeuing for this PVC.
+    1. Ensure PVC being deleted; to prevent inadvertent use by pods as we
+    change state to Secondary
+        1. If not being deleted, Continue to next PVC, but requeue VRG
+        1. **NOTE** The pods and the PVCs are expected to be not in use when
+        marking VR as Secondary, and hence are expected to be deleted before
+        we take action on the same
+        1. **NOTE** With OCM as the control plane, the VRG would be marked
+        `seconday` and then deleted, ensuring we get to a clean `secondary`
+        state before deleting the VRs for each PVC
+    1. Ensure PVC not [in use](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#storage-object-in-use-protection)
+        1. If in use, continue to next PVC, but requeue VRG
+    1. Update VR desired state for PVC as `secondary`
+1. For each PVC that we did not ignore above (IOW, all PVCs with vr-protected
+   finalizer)
+    1. Check if VR status has reached `secondary` state
+        1. If not, update status as progressing
+1. Report status as completed, if not progressing for any PVC
+
+## Finalizing VRG when deleted
+
+1. Skip PVCs that are not yet `vr-protected`
+1. When deleted as `primary`, delete the VR ensuring that it is `primary`,
+   ensure we update the PV `reclaimPolicy` to its original state, and remove the
+   PVC finalizer to aid PVC and PV garbage collection
+1. When deleted as `secondary`, delete the VR ensuring that it is `secondary`,
+   ensure we update the PV `reclaimPolicy` to its original, and remove the PVC
+   finalizer to aid PVC and PV garbage collection
+    1. **NOTE:** Expectation from CSI is that a delete of a `secondary` volume
+    is a success or a no-op, such that PVs are garbage collected on changing
+    the reclaim policy
+    1. **NOTE:** Further expectation from CSI is that a delete of a `primary`
+    volume would garbage collect the replicated volume as well
+
+- **TODO** When do the objects in the S3 store get cleaned up?
+- **TODO** How do we make the VRG restore from the S3 store, when created?
+      - This would be in lieu of the OCM hub taking restore actions, in a
+      non-OCM world (which is at present a user expectation)
```

Comment:
I suggested something like this to Raghu in yesterday's standup call.  We can add an optional VRG field, `operationMode` with values of (a) `backup-only` when managed by AVR on ACMHub & is the default mode, (b) `backup-and-restore`, when run independent of ACM, and (c) `backup-less`, when VRG is used only to control VRs.

Reply:
Ack! we need to have an option to do this. For now more an next phase entity.